Adds `FittingProblem` arg for model rebuild on evaluation, DFN examples and data-generation scripts

BradyPlanden opened the pull request:
# Description

Completes the following:

- Creates initial synthetic data generation examples
    - Aligns our examples to better represent true PyBOP usage 
    - Allows us to filter the `model.predict` output for discontinuities that occur from multi-step experiments (this currently causes large issues with the interpolated IDAKLU)
- Adds DFN identification examples


## Issue reference
Fixes #662 

## Review
Before you mark your PR as ready for review, please ensure that you've considered the following:
- Updated the [CHANGELOG.md](https://github.com/pybop-team/PyBOP/blob/develop/CHANGELOG.md) in reverse chronological order (newest at the top) with a concise description of the changes, including the PR number.
- Noted any breaking changes, including details on how it might impact existing functionality.

## Type of change
- [X] New Feature: A non-breaking change that adds new functionality.
- [ ] Optimization: A code change that improves performance.
- [X] Examples: A change to existing or additional examples.
- [ ] Bug Fix: A non-breaking change that addresses an issue.
- [ ] Documentation: Updates to documentation or new documentation for new features.
- [ ] Refactoring: Non-functional changes that improve the codebase.
- [ ] Style: Non-functional changes related to code style (formatting, naming, etc).
- [ ] Testing: Additional tests to improve coverage or confirm functionality.
- [ ] Other: (Insert description of change)

# Key checklist:

- [ ] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybop-team/PyBOP/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [ ] All unit tests pass: `$ nox -s tests`
- [ ] The documentation builds: `$ nox -s doctest`

You can run integration tests, unit tests, and doctests together at once, using `$ nox -s quick`.

## Further checks:
- [ ] Code is well-commented, especially in complex or unclear areas.
- [ ] Added tests that prove my fix is effective or that my feature works.
- [ ] Checked that coverage remains or improves, and added tests if necessary to maintain or increase coverage.

Thank you for contributing to our project! Your efforts help us to deliver great software.

martinjrobins reviewed (approved):
thanks @BradyPlanden, looks good to me :)

NicolaCourtier commented:
Thanks for these additions @BradyPlanden!

It seems to me that the safe option would be to set `build_on_evaluation` to `True` by default. But does this flag control rebuilding for geometric parameters or just rebuilding for an initial state?

BradyPlanden commented:
Thanks for the review/comments @martinjrobins @NicolaCourtier!

I've removed the extra `build_on_evaluation` argument and replaced it with the `model.set_initial_state` method in the examples. For the time being, I'd agree that this is more consistent with how we expect users to interact with the model and problem classes; however, I think we should reconsider allowing the management of the models' `initial_state` from both classes in the future, as it can be confusing to users how they differ, and when to use one versus the other. Anyway, I'll cross-post this to a separate issue for us to discuss!
